ui/pages/confirmations/components/confirm/info/typed-sign/permit-simulation/decoded-simulation/decoded-simulation.tsx

```diff
@@ -1,7 +1,95 @@
-import React from 'react';
+import React, { ReactNode } from 'react';
+import {
+  DecodingDataChangeType,
+  DecodingDataStateChange,
+} from '@metamask/signature-controller';
+import { Hex } from '@metamask/utils';
 
-const DecodedSimulation: React.FC<object> = () => (
-  <div>"DECODED SIMULATION IMPLEMENTATION"</div>
-);
+import { TokenStandard } from '../../../../../../../../../shared/constants/transaction';
+import {
+  Box,
+  Text,
+} from '../../../../../../../../components/component-library';
+import {
+  TextColor,
+  TextVariant,
+} from '../../../../../../../../helpers/constants/design-system';
+import { useI18nContext } from '../../../../../../../../hooks/useI18nContext';
+import { SignatureRequestType } from '../../../../../../types/confirm';
+import { useConfirmContext } from '../../../../../../context/confirm';
+import StaticSimulation from '../../../shared/static-simulation/static-simulation';
+import PermitSimulationValueDisplay from '../value-display/value-display';
+
+const getStateChangeLabelMap = (
+  t: ReturnType<typeof useI18nContext>,
+  changeType: string,
+) =>
+  ({
+    [DecodingDataChangeType.Transfer]: t('permitSimulationChange_transfer'),
+    [DecodingDataChangeType.Receive]: t('permitSimulationChange_receive'),
+    [DecodingDataChangeType.Approve]: t('permitSimulationChange_approve'),
+    [DecodingDataChangeType.Revoke]: t('permitSimulationChange_revoke'),
+    [DecodingDataChangeType.Bidding]: t('permitSimulationChange_bidding'),
+    [DecodingDataChangeType.Listing]: t('permitSimulationChange_listing'),
+  }[changeType]);
+
+const StateChangeRow = ({
+  stateChange,
+  chainId,
+}: {
+  stateChange: DecodingDataStateChange;
+  chainId: Hex;
+}) => {
+  const t = useI18nContext();
+  const { assetType, changeType, amount, contractAddress, tokenID } =
+    stateChange;
+  return (
+    <Box>
+      <Text
+        variant={TextVariant.bodyMdMedium}
+        color={TextColor.inherit}
+        marginLeft={2}
+      >
+        {getStateChangeLabelMap(t, changeType)}
+      </Text>
+      {(assetType === TokenStandard.ERC20 ||
+        assetType === TokenStandard.ERC721) && (
+        <PermitSimulationValueDisplay
+          tokenContract={contractAddress}
+          value={amount}
+          chainId={chainId}
+          tokenId={tokenID}
+          credit={changeType === DecodingDataChangeType.Receive}
+          debit={changeType === DecodingDataChangeType.Transfer}
+        />
+      )}
+    </Box>
+  );
+};
+
+const DecodedSimulation: React.FC<object> = () => {
+  const t = useI18nContext();
+  const { currentConfirmation } = useConfirmContext<SignatureRequestType>();
+  const chainId = currentConfirmation.chainId as Hex;
+  const { decodingLoading, decodingData } = currentConfirmation;
+
+  let stateChangeFragment: ReactNode[] = [];
+  if (decodingData?.stateChanges) {
```

Comment:
Minor, is there a need for this `if` and `let` as we could just `map` from the `stateChanges` with a `[]` fallback?

```
const stateChanges = decodingData?.stateChanges ?? [];
const stateChangeFragment = stateChanges.map(...);`
```

Reply:
fixed